src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Logging;
+using Microsoft.Azure.WebJobs.Script.Workers.Rpc;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Script.Workers
+{
+    internal class WorkerConcurrencyManager : IHostedService, IDisposable
+    {
+        private readonly TimeSpan _logStateInterval = TimeSpan.FromSeconds(60);
+        private readonly ILogger _logger;
+        private readonly IFunctionInvocationDispatcherFactory _functionInvocationDispatcherFactory;
+        private readonly IEnvironment _environment;
+
+        private WorkerConcurrencyOptions _workerConcurrencyOptions = new WorkerConcurrencyOptions();
+        private IFunctionInvocationDispatcher _functionInvocationDispatcher;
+        private System.Timers.Timer _timer;
+        private Stopwatch _addWorkerStopwatch = Stopwatch.StartNew();
+        private Stopwatch _logStateStopWatch = Stopwatch.StartNew();
+        private bool _disposed = false;
+
+        public WorkerConcurrencyManager(IFunctionInvocationDispatcherFactory functionInvocationDispatcherFactory,
+            IEnvironment environment, ILoggerFactory loggerFactory)
+        {
+            _functionInvocationDispatcherFactory = functionInvocationDispatcherFactory ?? throw new ArgumentNullException(nameof(functionInvocationDispatcherFactory));
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+
+            _logger = loggerFactory?.CreateLogger(LogCategories.Concurrency);
+        }
+
+        public WorkerConcurrencyOptions WorkerConcurrencyOptions
```

Comment:
Once you're taking from DI, you won't need this setter in tests so it can be removed.